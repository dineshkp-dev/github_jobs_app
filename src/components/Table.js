import React, { useState, useMemo } from 'react';
import TableData from './TableData';
import Pagination from './Pagination';
import { JOBS_PER_PAGE, SORT_ORDER } from '../common/constants';
import '../styles/table.css';

function Table(props) {
  const jobs = (props && props.jobs) || [];

  const [validJobTotal, setValidJobTotal] = useState(0);
  const [currentPage, setCurrentPage] = useState(1);
  const [searchText, setSearchText] = useState('');
  const [sortLocation, setSortLocation] = useState(false);
  const [sortOrder, setSortOrder] = useState(SORT_ORDER.ASC);

  const jobsData = useMemo(() => {
    let validJobs = jobs;
    if (searchText) {
      validJobs = validJobs.filter((job) =>
        job.location.toLowerCase().includes(searchText.toLowerCase())
      );
    }
    if (sortLocation) {
      validJobs = validJobs.sort((a, b) => {
        if (sortOrder === SORT_ORDER.ASC) {
          return a.location.localeCompare(b.location);
        } else {
          return b.location.localeCompare(a.location);
        }
      });
    }
    setValidJobTotal(validJobs.length);
    return validJobs.slice(
      (currentPage - 1) * JOBS_PER_PAGE,
      (currentPage - 1) * JOBS_PER_PAGE + JOBS_PER_PAGE
    );
  }, [jobs, currentPage, searchText, sortOrder, sortLocation]);

  const getSearchText = (e) => {
    const text = e.target.value;
    setSearchText(text);
    setCurrentPage(1);
  };

  return (
    <React.Fragment>
      <div className='table'>
        <div className='table-title'>
          <h1>Job details</h1>
        </div>
        <div id='table-header-container'>
          <Pagination
            total={validJobTotal}
            itemsPerPage={JOBS_PER_PAGE}
            currentPage={currentPage}
            onPageChange={(page) => setCurrentPage(page)}
          />
          <div className='search-container'>
            <input
              type='text'
              id='search-input'
              className='search-input'
              value={searchText}
              onChange={(e) => getSearchText(e)}
              placeholder='  Filter by job-location'
            />
          </div>
        </div>
        <table className='table-element'>
          <thead>
            <tr>
              <th>Type </th>
              <th>Title</th>
              <th
                onClick={(e) => {
                  setSortLocation(true);
                  sortOrder === SORT_ORDER.ASC
                    ? setSortOrder(SORT_ORDER.DESC)
                    : setSortOrder(SORT_ORDER.ASC);
                }}
                className='sortable-column'
              >
                Location{' '}
                {sortLocation ? (
                  sortOrder === SORT_ORDER.ASC ? (
                    <i className='fas fa-angle-up'></i>
                  ) : (
                    <i className='fas fa-angle-down'></i>
                  )
                ) : null}
              </th>
            </tr>
          </thead>
          <tbody>
            {jobsData && jobsData.length ? (
              jobsData.map((job) => <TableData key={job.id} jobdetail={job} />)
            ) : (
              <tr>
                <td colSpan='10'>
                  <div className='no-data-available'>No Data is available</div>
                </td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    </React.Fragment>
  );
}

export default Table;
