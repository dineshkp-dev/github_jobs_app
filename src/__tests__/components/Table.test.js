import Table from '../../components/Table';
import { jobs, multiJobs, multiJobsMultiPage } from '../utils/testUtils';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Table snapshot test', () => {
  test('renders', () => {
    const { container } = render(
      <Router>
        <Table jobs={jobs} />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Table pagination test', () => {
  test('pagination test', async () => {
    render(
      <Router>
        <Table jobs={multiJobsMultiPage} />
      </Router>
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('Type')).toBeInTheDocument();
    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Location')).toBeInTheDocument();
    expect(screen.getAllByText('Full Time')).toHaveLength(20);
    expect(screen.getAllByText('Full Stack Engineer')).toHaveLength(20);
    expect(screen.getByText('AAA')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Filter by job-location')
    ).toBeInTheDocument();
    const nextButton = screen.getByText('Next');
    fireEvent.click(nextButton);
    await waitFor(() => {
      expect(screen.getByText('ZZZ')).toBeInTheDocument();
    });
  });
});

describe('Table render test', () => {
  test('renders Table', () => {
    render(
      <Router>
        <Table jobs={jobs} />
      </Router>
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('Type')).toBeInTheDocument();
    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Location')).toBeInTheDocument();
    expect(screen.getByText('Full Time')).toBeInTheDocument();
    expect(screen.getByText('Full Stack Engineer')).toBeInTheDocument();
    expect(screen.getByText('New York City')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Filter by job-location')
    ).toBeInTheDocument();
  });

  test('when no props are passed', () => {
    render(
      <Router>
        <Table />
      </Router>
    );
    expect(screen.getByText('Type')).toBeInTheDocument();
    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Location')).toBeInTheDocument();
    expect(screen.getByText('No Data is available')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Filter by job-location')
    ).toBeInTheDocument();
    screen.debug();
  });

  test('renders Table when jobs is empty', () => {
    const emptyJobs = [];
    render(
      <Router>
        <Table jobs={emptyJobs} />
      </Router>
    );
    expect(screen.getByText('Type')).toBeInTheDocument();
    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Location')).toBeInTheDocument();
    expect(screen.getByText('No Data is available')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Filter by job-location')
    ).toBeInTheDocument();
    screen.debug();
  });

  test('renders Table when jobs has multiple values', () => {
    render(
      <Router>
        <Table jobs={multiJobs} />
      </Router>
    );
    expect(screen.getByText('Previous')).toBeInTheDocument();
    expect(screen.getByText('Next')).toBeInTheDocument();
    expect(screen.getByText('1')).toBeInTheDocument();
    expect(screen.getByText('Type')).toBeInTheDocument();
    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Location')).toBeInTheDocument();
    expect(screen.getAllByText('Full Time')).toHaveLength(4);
    expect(screen.getAllByText('Full Stack Engineer')).toHaveLength(4);
    expect(screen.getByText('AAA')).toBeInTheDocument();
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
    expect(screen.getByText('DDD')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Filter by job-location')
    ).toBeInTheDocument();
  });
});

describe('Filtering in table test', () => {
  test('filter table content', async () => {
    render(
      <Router>
        <Table jobs={multiJobs} />
      </Router>
    );
    expect(screen.getByText('AAA')).toBeInTheDocument();
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
    const inputField = screen.getByPlaceholderText('Filter by job-location');
    fireEvent.change(inputField, { target: { value: 'AAA' } });
    await waitFor(() => {
      expect(screen.getByText('AAA')).toBeInTheDocument();
    });
    expect(screen.queryByText('BBB')).toBeNull();
    expect(screen.queryByText('CCC')).toBeNull();
  });

  test('filter table content and then show all', async () => {
    render(
      <Router>
        <Table jobs={multiJobs} />
      </Router>
    );
    expect(screen.getByText('AAA')).toBeInTheDocument();
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
    const inputField = screen.getByPlaceholderText('Filter by job-location');
    fireEvent.change(inputField, { target: { value: 'AAA' } });
    await waitFor(() => {
      expect(screen.getByText('AAA')).toBeInTheDocument();
    });
    expect(screen.queryByText('BBB')).toBeNull();
    expect(screen.queryByText('CCC')).toBeNull();
    fireEvent.change(inputField, { target: { value: '' } });
    await waitFor(() => {
      expect(screen.getByText('AAA')).toBeInTheDocument();
    });
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
  });
});

describe('Sorting in table test', () => {
  test('sorting table content', async () => {
    render(
      <Router>
        <Table jobs={multiJobs} />
      </Router>
    );
    expect(screen.getByText('AAA')).toBeInTheDocument();
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
    expect(screen.getByText('DDD')).toBeInTheDocument();
    const locationsColumn = screen.getByText('Location');
    fireEvent.click(locationsColumn);
    await waitFor(() => {
      expect(screen.getByText('AAA')).toBeInTheDocument();
    });
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
    expect(screen.getByText('DDD')).toBeInTheDocument();
    fireEvent.click(locationsColumn);
    await waitFor(() => {
      expect(screen.getByText('AAA')).toBeInTheDocument();
    });
    expect(screen.getByText('BBB')).toBeInTheDocument();
    expect(screen.getByText('CCC')).toBeInTheDocument();
    expect(screen.getByText('DDD')).toBeInTheDocument();
  });
});
