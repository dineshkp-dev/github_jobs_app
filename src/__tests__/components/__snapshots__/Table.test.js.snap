// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table snapshot test renders 1`] = `
<div>
  <div
    class="table"
  >
    <div
      class="table-title"
    >
      <h1>
        Job details
      </h1>
    </div>
    <div
      id="table-header-container"
    >
      <div
        class="pagination-container"
      >
        <div
          class="button-disabled"
        >
          <div
            class="button-text"
          >
            Previous
          </div>
        </div>
        <div
          class="pagination-class-indicator"
        >
          <div
            class="pagination-page active-page"
          >
            1
          </div>
        </div>
        <div
          class="button-disabled"
        >
          <div
            class="button-text"
          >
            Next
          </div>
        </div>
      </div>
      <div
        class="search-container"
      >
        <input
          class="search-input"
          id="search-input"
          placeholder="  Filter by job-location"
          type="text"
          value=""
        />
      </div>
    </div>
    <table
      class="table-element"
    >
      <thead>
        <tr>
          <th>
            Type 
          </th>
          <th>
            Title
          </th>
          <th
            class="sortable-column"
          >
            Location
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Full Time
          </td>
          <td>
            <a
              href="/jobdetails/testid"
            >
              Full Stack Engineer
            </a>
          </td>
          <td>
            New York City
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
